File: app/src/main/java/com/github/ymegane/android/taptaptap/presentation/view/TapView.java
package com.github.ymegane.android.taptaptap.presentation.view;

import android.content.Context;
import android.os.Vibrator;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.widget.FrameLayout;

import com.jakewharton.rxbinding.view.RxView;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.Observer;
import rx.Subscription;
import rx.functions.Action1;
import rx.functions.Func1;

import com.github.ymegane.android.taptaptap.presentation.view.CircleView.Circle;

public class TapView extends FrameLayout {

    private OnTapListener tapListener;

    public TapView(Context context) {
        super(context);
        init();
    }

    public TapView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public TapView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public TapView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    public void setOnTapListener(OnTapListener listener) {
        this.tapListener = listener;
    }

    private void init() {
        prepareForTouchEvent();
    }

    private long lastVibratedTime;
    private Subscription subscription;
    private void prepareForTouchEvent() {
        subscription = RxView.touches(this).flatMap(new Func1<MotionEvent, Observable<Circle>>() {
            @Override
            public Observable<Circle> call(MotionEvent event) {
                int n = event.getPointerCount();

                List<Circle> circles = new ArrayList<>(n);
                for (int i=0; i<n; i++) {
                    Circle circle = new Circle(i, event);
                    circles.add(circle);
                }
                return Observable.from(circles);
            }
        }).doOnNext(new Action1<Circle>() {
            @Override
            public void call(Circle circle) {
                MotionEvent event = circle.event;
                if (event.getAction() == MotionEvent.ACTION_MOVE) {
                    if (event.getEventTime() - lastVibratedTime > 600) {
                        lastVibratedTime = event.getEventTime();
                        vibrate();
                    }
                    CircleView view = findCircleView(circle);
                    if (view != null) {
                        view.continueCancelTimer();
                    }
                }
            }
        }).doOnNext(new Action1<Circle>() {
            @Override
            public void call(Circle circle) {
                MotionEvent event = circle.event;
                if (event.getActionMasked() == MotionEvent.ACTION_UP
                        || event.getActionMasked() == MotionEvent.ACTION_POINTER_UP
                        || event.getActionMasked() == MotionEvent.ACTION_CANCEL) {
                    CircleView view = findCircleView(circle);
                    if (view != null) {
                        view.startAnimation();
                    }
                }
            }
        }).filter(new Func1<Circle, Boolean>() {
            @Override
            public Boolean call(Circle circle) {
                MotionEvent event = circle.event;
                return event.getActionMasked() == MotionEvent.ACTION_DOWN ||
                        event.getActionMasked() == MotionEvent.ACTION_POINTER_DOWN;
            }
        }).filter(new Func1<Circle, Boolean>() {
            @Override
            public Boolean call(Circle circle) {
                return findCircleView(circle) == null;
            }
        })
        .filter(new Func1<Circle, Boolean>() {
            @Override
            public Boolean call(Circle circle) {
                return getChildCount() < 100;
            }
        })
        .subscribe(new Observer<Circle>() {
            @Override
            public void onCompleted() {
            }

            @Override
            public void onError(Throwable e) {
            }

            @Override
            public void onNext(Circle circle) {
                addCircleView(circle);
                vibratePattern();
                if (tapListener != null) {
                    tapListener.onTap();
                }
            }
        });
    }

    public void release() {
        if (subscription != null) {
            subscription.unsubscribe();
        }
    }

    @Nullable
    private CircleView findCircleView(Circle circle) {
        int n = getChildCount();
        for (int i=0; i < n; i++) {
            CircleView child = (CircleView) getChildAt(i);
            if (child.equals(circle)) {
                return child;
            }
        }
        return null;
    }

    private void addCircleView(Circle circle) {
        CircleView view = new CircleView(getContext(), circle);
        addView(view);
        view.invalidate();
    }

    private void vibrate() {
        Vibrator vibrator = (Vibrator) getContext().getSystemService(Context.VIBRATOR_SERVICE);
        if (!vibrator.hasVibrator()) {
            return;
        }

        vibrator.vibrate(500);
    }

    private static final long[] PATTERN = new long[]{100, 500};

    private void vibratePattern() {
        Vibrator vibrator = (Vibrator) getContext().getSystemService(Context.VIBRATOR_SERVICE);
        if (!vibrator.hasVibrator()) {
            return;
        }

        vibrator.vibrate(PATTERN, 1);
    }

    public interface OnTapListener {
        void onTap();
    }
}

File: app/src/main/java/com/github/ymegane/android/taptaptap/presentation/view/TapView.kt
package com.github.ymegane.android.taptaptap.presentation.view

import android.content.Context
import android.os.Vibrator
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.FrameLayout
import com.github.ymegane.android.taptaptap.domain.model.Circle

import com.jakewharton.rxbinding.view.RxView

import java.util.ArrayList

import rx.Observable
import rx.Subscription

class TapView : FrameLayout {

    var tapListener: OnTapListener? = null

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        prepareForTouchEvent()
    }

    private var lastVibratedTime: Long = 0
    private var subscription: Subscription? = null
    private fun prepareForTouchEvent() {
        subscription = RxView.touches(this).flatMap { event ->
            val n = event.pointerCount

            // Add circles for the number of points
            val circles = ArrayList<Circle>(n)
            (0..n - 1).mapTo(circles) { Circle(it, event) }
            Observable.from(circles)
        }.doOnNext { circle ->
            val event = circle.event
            if (event.action == MotionEvent.ACTION_MOVE) {
                // If ViewGroup already has circles, continue to display
                if (event.eventTime - lastVibratedTime > 600) {
                    lastVibratedTime = event.eventTime
                    vibrate()
                }
                val view = findCircleView(circle)
                view?.continueCancelTimer()
            } else if (event.actionMasked == MotionEvent.ACTION_UP
                    || event.actionMasked == MotionEvent.ACTION_POINTER_UP
                    || event.actionMasked == MotionEvent.ACTION_CANCEL) {
                // If ViewGroup already has circles, start hiding animation
                val view = findCircleView(circle)
                view?.startAnimation()
            }
        }.filter { circle ->
            // Filter down event
            val event = circle.event
            event.actionMasked == MotionEvent.ACTION_DOWN || event.actionMasked == MotionEvent.ACTION_POINTER_DOWN
        }.filter { circle -> findCircleView(circle) == null }
        .filter { childCount < 100 }
        .subscribe { circle ->
            addCircleView(circle)
            vibratePattern()
            tapListener?.onTap()
        }
    }

    fun release() = subscription?.unsubscribe()

    private fun findCircleView(circle: Circle): CircleView? {
        val n = childCount
        return (0..n - 1)
                .map { getChildAt(it) as CircleView }
                .firstOrNull { it.equals(circle) }
    }

    private fun addCircleView(circle: Circle) {
        val view = CircleView(context, circle)
        addView(view)
        view.invalidate()
    }

    private fun vibrate() {
        val vibrator = context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        if (!vibrator.hasVibrator()) {
            return
        }

        vibrator.vibrate(500)
    }

    private fun vibratePattern() {
        val vibrator = context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        if (!vibrator.hasVibrator()) {
            return
        }

        vibrator.vibrate(PATTERN, 1)
    }

    interface OnTapListener {
        fun onTap()
    }

    companion object {
        private val PATTERN = longArrayOf(100, 500)
    }
}
